plugin-parser/mlops-plugin-parser.cc

```diff
@@ -88,10 +88,128 @@ typedef enum {
 } mlsvc_json_type_e;
 
 /**
- * @brief Parse json and update ml-service database via invoking daemon.
+ * @brief Internal enumeration for package manager event types.
+ */
+typedef enum {
+  MLSVC_PKGMGR_MDPARSER_PLUGIN_EVENT_TYPE_INSTALL = 0,
+  MLSVC_PKGMGR_MDPARSER_PLUGIN_EVENT_TYPE_UNINSTALL = 1,
+  MLSVC_PKGMGR_MDPARSER_PLUGIN_EVENT_TYPE_UPGRADE = 2,
+  MLSVC_PKGMGR_MDPARSER_PLUGIN_EVENT_TYPE_MAX
+} mlsvc_package_manager_event_type_e;
+
+/**
+ * @brief Internal function for uninstall models and pipelines.
  */
 static void
-_parse_json (JsonNode *node, mlsvc_json_type_e json_type, const gchar *app_info)
+_uninstall_rpk (const char *name, const char *_info, mlsvc_json_type_e json_type)
+{
+  g_autoptr (JsonParser) parser = NULL;
+  g_autoptr (GError) err = NULL;
+  JsonNode *root;
+  JsonObject *object;
+  JsonArray *array = NULL;
+
+  parser = json_parser_new ();
+  guint _info_len = 1U;
+
+  if (!json_parser_load_from_data (parser, _info, -1, &err)) {
+    _E ("Failed to load '%s' info data.", name);
+    return;
+  }
+
+  root = json_parser_get_root (parser);
+  if (!root) {
+    _E ("Failed to load '%s' info, cannot get the top node from info.", name);
+    return;
+  }
+
+  if (JSON_NODE_HOLDS_ARRAY (root)) {
+    array = json_node_get_array (root);
+    if (!array) {
+      _E ("Failed to get root array from '%s' info", name);
+      return;
+    }
+
+    _info_len = json_array_get_length (array);
+  }
+
+  /* Update ML service database. */
+  for (guint i = 0; i < _info_len; ++i) {
+    int ret = 0;
+    if (array)
+      object = json_array_get_object_element (array, i);
+    else
+      object = json_node_get_object (root);
+
+    const gchar *app_info = json_object_get_string_member (object, "app_info");
```

Comment:
suggestion:
we cannot sure the model or resource with 'name' always have app_info field. (If user registers new model with 'name' in user's application, it does not have app_info data.)

```
const gchar *app_info = json_object_get_string_member (object, "app_info");
/* If app info is empty string, it is not installed from rpk. */
if (!STR_IS_VALID (app_info)
  continue;
```

Reply:
thanks, i update it.